### PR TITLE
Add dynamic uniform and storage buffer support

### PR DESCRIPTION
**Pull request by asuonpaa**

### CODE REVIEW

File: tests/cases/graphics_dynamic_buffers.amber [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+layout(location = 0) in vec4 frag_color;
+layout(location = 0) out vec4 final_color;
+
+void main()
+{
+    final_color = frag_color;
+}
+END
+
+# Meaningful data is read using minimum offset alignment of 256 bytes.
```
Comment (paulthomson) [inline tests/cases/graphics_dynamic_buffers.amber:64]: Similar to above comment: if this is due to the Vulkan spec, please say so. E.g. The Vulkan spec guarantees that the min offset alignment required is at most 256, and must be a power of two (probably...the spec is not 100% clear). Thus, aligning offsets to 256 should always be valid.

File: docs/amber_script.md [hunk 2]
```diff
@@ -550,7 +552,8 @@ All BIND BUFFER and BIND SAMPLER commands below define a descriptor set and bind
 These commands can be replaced with BIND BUFFER_ARRAY and BIND SAMPLER_ARRAY commands.
 In these cases multiple buffer or sampler names need to be provided, separated by spaces.
 This creates a descriptor array of buffers or samplers bound to the same descriptor set
-and binding ID.
+and binding ID. An array of offsets should be provided for OFFSET command when using
+dynamic buffers with BUFFER_ARRAY.
```
[the remark this answers is not in the document]
Comment (asuonpaa) [reply docs/amber_script.md:556]: done

File: src/amberscript/parser.cc [hunk 3]
```diff
@@ -1128,10 +1134,32 @@ Result Parser::ParsePipelineBind(Pipeline* pipeline) {
...
+
+          for (size_t i = 0; i < buffers.size(); i++) {
+            token = tokenizer_->NextToken();
+
+            if (!token->IsInteger()) {
+              if (i > 0)
+                return Result(
+                    "expecting an OFFSET value for each buffer in the array");
+              else
+                return Result("expecting an integer value for OFFSET");
```
[the remark this answers is not in the document]
Comment (asuonpaa) [reply src/amberscript/parser.cc:1152]: done

File: src/vulkan/pipeline.cc [hunk 4]
```diff
@@ -337,6 +342,10 @@ Result Pipeline::AddBufferDescriptor(const BufferCommand* cmd) {
         cmd->GetBuffer());
   }
 
+  if (cmd->IsUniformDynamic() || cmd->IsSSBODynamic())
+    reinterpret_cast<BufferDescriptor*>(desc)->AddDynamicOffset(
+        cmd->GetDynamicOffset());
+
```
[the remark this answers is not in the document]
Comment (asuonpaa) [reply src/vulkan/pipeline.cc:348]: done

File: tests/cases/compute_dynamic_buffers.amber [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+BUFFER buf0 DATA_TYPE vec4<float> SIZE 17 FILL 0.0
+BUFFER buf1 DATA_TYPE vec4<float> SIZE 17 FILL 0.0
+
+PIPELINE compute pipeline
+  ATTACH compute_shader
+
+  BIND BUFFER buf0 AS storage_dynamic DESCRIPTOR_SET 0 BINDING 0 OFFSET 0
+  BIND BUFFER buf1 AS storage_dynamic DESCRIPTOR_SET 0 BINDING 1 OFFSET 256
+END
+
```
[the remark this answers is not in the document]
Comment (asuonpaa) [reply tests/cases/compute_dynamic_buffers.amber:45]: As discussed earlier that needs refactoring of Vulkan engine. I'd rather do that as a separate PR.

File: tests/cases/compute_dynamic_buffers.amber [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+void main()
+{
+    data0 = vec4(1, 2, 3, 4);
+    data1 = vec4(5, 6, 7, 8);
+}
+END
+
+BUFFER buf0 DATA_TYPE vec4<float> SIZE 17 FILL 0.0
+BUFFER buf1 DATA_TYPE vec4<float> SIZE 17 FILL 0.0
```
[the remark this answers is not in the document]
Comment (asuonpaa) [reply tests/cases/compute_dynamic_buffers.amber:37]: done

File: tests/cases/compute_dynamic_buffers.amber [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+PIPELINE compute pipeline
+  ATTACH compute_shader
+
+  BIND BUFFER buf0 AS storage_dynamic DESCRIPTOR_SET 0 BINDING 0 OFFSET 0
+  BIND BUFFER buf1 AS storage_dynamic DESCRIPTOR_SET 0 BINDING 1 OFFSET 256
```
[the remark this answers is not in the document]
Comment (asuonpaa) [reply tests/cases/compute_dynamic_buffers.amber:43]: done

File: tests/cases/graphics_dynamic_buffers.amber (hunk 1 above)
Comment (asuonpaa) [reply tests/cases/graphics_dynamic_buffers.amber:64]: done

**Comment by paulthomson** [comment]
The CI previously passed. Only comments have changed since. It looks like two configs are stuck or timing out, so I will merge.

**Pull request merged**